src/pgo/model/distsys/CentralizedEtcdStateStrategy.java

```diff
@@ -0,0 +1,295 @@
+package pgo.model.distsys;
+
+import pgo.PGoNetOptions;
+import pgo.model.golang.*;
+import pgo.model.intermediate.*;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Vector;
+import java.util.stream.Stream;
+
+public class CentralizedEtcdStateStrategy implements StateStrategy {
+	public static final String GLOBAL_STATE_OBJECT = "globalState";
+
+	private PGoNetOptions.StateOptions stateOptions;
+
+	public CentralizedEtcdStateStrategy(PGoNetOptions.StateOptions stateOptions) {
+		this.stateOptions = stateOptions;
+	}
+
+	@Override
+	public void generateConfig(GoProgram go) {
+		Vector<Statement> topLevelMain = go.getMain().getBody();
+		String configObj = "cfg";
+
+		Assignment cfgDecl = new Assignment(
+				new Vector<>(Collections.singletonList(configObj)),
+				new Expression() {
+					@Override
+					public Vector<String> toGo() {
+						StructDefinition sdef = new StructDefinition("distsys.Config", true);
+						sdef.addField("Endpoints", new Expression() {
+							@Override
+							public Vector<String> toGo() {
+								Vector<String> endpoints = new Vector<>();
+								for (String h : stateOptions.endpoints) {
+									endpoints.add(String.format("\"http://%s\"", h));
+								}
+
+								return new Token(String.format("%s{%s}",
+										new PGoCollectionType.PGoSlice("string").toGo(),
+										String.join(", ", endpoints))).toGo();
+							}
+						});
+						sdef.addField("Timeout", new Token(Integer.toString(stateOptions.timeout)));
+						return sdef.toGo();
+					}
+				},
+				true
+		);
+		topLevelMain.add(cfgDecl);
+
+		VariableDeclaration errDecl = new VariableDeclaration(
+				"err", PGoType.inferFromGoTypeName("error"), null, false, false, false
+		);
+		topLevelMain.add(errDecl);
+
+		Assignment stateObj = new Assignment(
+				new Vector<>(Arrays.asList(GLOBAL_STATE_OBJECT, "err")),
+				new FunctionCall("distsys.InitEtcdState",
+						new Vector<>(Collections.singletonList(new Token(configObj)))),
+				false
+		);
+		topLevelMain.add(stateObj);
+
+		go.getImports().addImport("os");
+		Vector<Statement> ifBody = new Vector<Statement>() {
+			{
+				add(new Comment("handle error - could not connect to etcd", false));
+				add(new FunctionCall("os.Exit",
+						new Vector<>(Collections.singletonList(new Token("1")))));
+			}
+		};
+
+		pgo.model.golang.If errIf = new pgo.model.golang.If(new Token("err != nil"), ifBody, new Vector<>());
+		topLevelMain.add(errIf);
+	}
+
+	@Override
+	public void generateGlobalVariables(GoProgram go) {
+		VariableDeclaration stateDecl = new VariableDeclaration(GLOBAL_STATE_OBJECT,
+				new PGoMiscellaneousType.EtcdState(), null, false, false, false);
+
+		go.addGlobal(stateDecl);
+	}
+
+	@Override
+	public void initializeGlobalVariables(GoProgram go) {
```

Comment:
Perhaps a better name (the one we've been using when we talk about it) would be `initializeGlobalState`.

Reply:
agreed